src/Microsoft.ML.TorchSharp/Utils/ImageUtils.cs

```diff
@@ -0,0 +1,221 @@
+                        boxes[boxIndex++] = bbox[1].ToSingle();
+                        boxes[boxIndex++] = (bbox[2] - bbox[0] + 1).ToSingle();
+                        boxes[boxIndex++] = (bbox[3] - bbox[1] + 1).ToSingle();
+                    }
+                }
+            }
+        }
+
+        private static Tensor Nms(Tensor boxes, Tensor scores, double iouThreshold = 0.5)
+        {
+            using (var nmsScope = torch.NewDisposeScope())
+            {
+                // boxes: Tensor [N,4]，scores: Tensor [N,]
+                var x1 = boxes[.., 0];
+                var y1 = boxes[.., 1];
+                var x2 = boxes[.., 2];
+                var y2 = boxes[.., 3];
+                var areas = (x2 - x1) * (y2 - y1); // [N,]
+
+                var (_, _order) = scores.sort(0, descending: true);
+
+                var keep = new List<long>();
+                var order = _order[..];
+                while (order.numel() > 0)
+                {
+                    long i;
+                    if (order.numel() == 1)
+                    {
+                        i = order.cpu().item<long>();
+                        keep.Add(i);
+                        break;
+                    }
+                    else
+                    {
+                        i = order[0].cpu().item<long>();
+                        keep.Add(i);
+                    }
+
+                    var xx1 = x1[order[1..]].clamp(min: x1[i]); // [N - 1,]
+                    var yy1 = y1[order[1..]].clamp(min: y1[i]);
+                    var xx2 = x2[order[1..]].clamp(max: x2[i]);
+                    var yy2 = y2[order[1..]].clamp(max: y2[i]);
+                    var inter = (xx2 - xx1).clamp(min: 0) * (yy2 - yy1).clamp(min: 0); // [N - 1,]
+
+                    var iou = inter / (areas[i] + areas[order[1..]] - inter); // [N-1, ]
+                    var idx = (iou <= iouThreshold).nonzero().squeeze(); // idx: [N - 1,] and order:[N,]
+                    if (idx.numel() == 0)
+                    {
+                        break;
+                    }
+
+                    order = order[idx + 1];
+                }
+
+                var ids = torch.from_array(keep.ToArray()).to_type(ScalarType.Int64).to(device: boxes.device);
+                return ids.MoveToOuterDisposeScope();
+            }
+        }
+
+        /// <summary>
+        /// Transform bounding boxes with delta.
+        /// </summary>
+        /// <param name="boxes">The bounding boxes to be transformed.</param>
+        /// <param name="deltas">The deltas in transform.</param>
+        /// <returns>The transformed boundbing boxes.</returns>
+        private static Tensor TransformBbox(Tensor boxes, Tensor deltas)
+        {
+            using (var transformBboxScope = torch.NewDisposeScope())
+            {
+                var mean = torch.from_array(new double[] { 0, 0, 0, 0 }).to_type(ScalarType.Float32).to(boxes.device);
```

Comment:
How's these mean and std get from and why we need to denormalize the regression

Reply:
Honestly not really sure. This is the 100% a copy of the code I was given with the model in order to interpret the bounding boxes. It doesn't look to me like it will do anything, but I didn't want ot mess with it so I'm going to leave it for now as is.